lib/sbi_ipi.c

```diff
@@ -23,6 +23,7 @@ static int sbi_ipi_send(struct sbi_scratch *scratch, u32 hartid, u32 event)
 {
 	struct sbi_scratch *remote_scratch = NULL;
 	const struct sbi_platform *plat = sbi_platform_ptr(scratch);
+	unsigned long tmp = remote_scratch->ipi_type;
```

Comment:
For sure, this breaks atomicity of "ipi_type". The "tmp" variable is on stack so it's location will change based on calling path.

We have three options here:
1. Drop __packed from "struct sbi_scratch" since all members are "unsigned long". May be some compiler expert can clarify if this would work fine.
2. Change "ipi_type" to "ipi_data" (pointer) which will point to per-HART "struct sbi_ipi_data" instance. The "struct sbi_ipi_data" will have "unsigned long ipi_type" member.
3. We have ensured that most packed structures have members with appropriate alignment so we suppress the GCC9 warning using some GCC9 option.

Reply:
Yep, you are right. This definitely does not work. Let me check this again tomorrow.

Reply:
Why don't we just change ipi_type to be a pointer to an unsigned long? It's never set in assembly only ever by the atomics.